docs/benchmarkdotnet.md

```diff
@@ -0,0 +1,332 @@
+# BenchmarkDotNet
+
+BenchmarkDotNet is the benchmarking tool that allows to run benchmarks for .NET, .NET Core, CoreRT and Mono. Including private runtime builds.
+
+## Table of Contents
+
+- [Main Concepts](#Main-Concepts)
+- [Running the Benchmarks](#Running-the-Benchmarks)
+  - [Prerequisites](#Prerequisites)
+  - [Interactive Mode](#Interactive-Mode)
+  - [Command Line](#Command-Line)
+    - [Filtering the Benchmarks](#Filtering-the-Benchmarks)
+    - [Listing the Benchmarks](#Listing-the-Benchmarks)
+- [Reading the Results](#Reading-the-Results)
+  - [Reading the Histogram](#Reading-the-Histogram)
+  - [Reading Memory Statistics](#Reading-Memory-Statistics)
+- [Multiple Runtimes](#Multiple-Runtimes)
+- [Regressions](#Regressions)
+- [Profiling](#Profiling)
+- [Disassembly](#Disassembly)
+- [Private Runtime Builds](#Private-Runtime-Builds)
+  - [Running In Process](#Running-In-Process)
+  - [CoreRun](#CoreRun)
+  - [dotnet cli](#dotnet-cli)
+  - [Private CLR Build](#Private-CLR-Build)
+  - [Private CoreRT Build](#Private-CoreRT-Build)
+
+
+## Main Concepts
+
+Benchmarking is really hard (especially microbenchmarking), you can easily make a mistake during performance measurements.
+BenchmarkDotNet will protect you from the common pitfalls (even for experienced developers) because it does all the dirty work for you:
+
+* it generates an isolated project per runtime (with boilerplate code)
+* it builds the project in `Release` (using Roslyn or dotnet cli)
+* it runs every benchmark in a stand-alone process (to achieve process isolation and avoid side effects)
+* it estimates the perfect invocation count per iteration (based on `IterationTime`)
+* it warms-up the code
+* it evaluates the overhead
+* it runs multiple iterations of the method until the requested level of precision is met
+* it consumes the benchmark result to avoid dead code elimination
+* it prevents from inlining of the benchmark by wrapping it with a delegate
+* it prints the results to the console in GitHub markdown, so you can just copy-paste the printed table to GitHub
+* it exports the results to `.\BenchmarkDotNet.Artifacts\results` so you can store them for later use.
```

Comment:
nit: change this `.\BenchmarkDotNet.Artifacts\results` to this `BenchmarkDotNet.Artifacts\results`

Reply:
why? .\ is very reasonable there.

Reply:
Why is it reasonable? What am I missing? `.\` implies the current directory, which is not always the case. After you build/run, regardless of the working directory, the output folder "BenchmarkDotNet.Artifacts\results" will be under something like this (from the root of the repo):
`.\artifacts\bin\MicroBenchmarks\{Optionally the architecture of the Cli}\Release\{Target Framework}\BenchmarkDotNet.Artifacts\results`.

Reply:
Interesting. That hasn't been my experience. 
```cmd
C:\Users\billw\source\repos\bdntest> dir


    Directory: C:\Users\billw\source\repos\bdntest


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        2/15/2019   2:52 PM                .vscode
d-----        2/15/2019   2:53 PM                bin
d-----        2/15/2019   2:53 PM                obj
-a----        2/15/2019   2:50 PM            280 bdntest.csproj
-a----        2/15/2019   2:57 PM            368 Program.cs


C:\Users\billw\source\repos\bdntest> dotnet run -c Release | Out-Null
C:\Users\billw\source\repos\bdntest> dir


    Directory: C:\Users\billw\source\repos\bdntest


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        2/15/2019   2:52 PM                .vscode
d-----        2/15/2019   2:57 PM                BenchmarkDotNet.Artifacts
d-----        2/15/2019   2:53 PM                bin
d-----        2/15/2019   2:53 PM                obj
-a----        2/15/2019   2:50 PM            280 bdntest.csproj
-a----        2/15/2019   2:57 PM            368 Program.cs

```

The Microbenchmark project is different but as this is the documentation stub about BenchmarkDotNet (and its default behavior is as it is) I think this is fine.

Reply:
[this](https://github.com/dotnet/performance/blob/master/src/benchmarks/micro/Program.cs#L32) is where we override the default. I'm not sure why we do this honestly; burying the artifacts like this I have found personally frustrating.

Reply:
I think BenchmarkDotNet did not take into account generating the artifacts for multiple runs for competitive analysis. If you do not keep it with the target framework you ran, then it's even more frustrating figuring this out. Your example is the default BenchmarkDotNet output for a hello work kind of project, but not the default for this repo. Personally, I feel that BDN should output to the bin directory and not side-by-side the sources.

Reply:
I also feel that this is documentation about the BDN harnessed benchmarks in the repo, not about BDN.

Reply:
I created an issue in BDN for that: https://github.com/dotnet/BenchmarkDotNet/issues/1074